File: Sources/xlsOneCore/CompoundFileReader.swift
import Foundation

enum CompoundFileError: Error, CustomStringConvertible {
    case invalidHeader
    case unsupportedSectorSize(Int)
    case chainLoop(Int)
    case sectorOutOfBounds(Int)
    case streamNotFound([String])

    var description: String {
        switch self {
        case .invalidHeader:
            return "不是有效的 OLE Compound File。"
        case .unsupportedSectorSize(let size):
            return "暂不支持的 OLE sector size: \(size)。"
        case .chainLoop(let sector):
            return "OLE sector chain 出现循环: \(sector)。"
        case .sectorOutOfBounds(let sector):
            return "OLE sector 越界: \(sector)。"
        case .streamNotFound(let names):
            return "未找到工作簿数据流: \(names.joined(separator: ", "))。"
        }
    }
}

struct CompoundFileReader {
    private struct DirectoryEntry {
        let name: String
        let type: UInt8
        let startingSector: Int
        let streamSize: Int
    }

    private static let endOfChain = UInt32(0xFFFFFFFE)
    private static let freeSector = UInt32(0xFFFFFFFF)
    private static let difatSector = UInt32(0xFFFFFFFC)
    private static let fatSector = UInt32(0xFFFFFFFD)

    private let data: Data
    private let sectorSize: Int
    private let miniSectorSize: Int
    private let miniStreamCutoffSize: Int
    private let firstDirectorySector: Int
    private let firstMiniFATSector: Int
    private let numberOfMiniFATSectors: Int
    private let fat: [UInt32]
    private let miniFAT: [UInt32]
    private let directoryEntries: [DirectoryEntry]
    private let rootEntry: DirectoryEntry?
    private let miniStream: Data

    init(data: Data) throws {
        guard data.count >= 512,
              data.prefix(8).elementsEqual([0xD0, 0xCF, 0x11, 0xE0, 0xA1, 0xB1, 0x1A, 0xE1]) else {
            throw CompoundFileError.invalidHeader
        }

        self.data = data

        let sectorShift = Int(data.xlsUInt16(at: 30))
        let miniSectorShift = Int(data.xlsUInt16(at: 32))
        self.sectorSize = 1 << sectorShift
        self.miniSectorSize = 1 << miniSectorShift
        self.firstDirectorySector = Int(data.xlsUInt32(at: 48))
        self.miniStreamCutoffSize = Int(data.xlsUInt32(at: 56))
        self.firstMiniFATSector = Int(data.xlsUInt32(at: 60))
        self.numberOfMiniFATSectors = Int(data.xlsUInt32(at: 64))

        guard sectorSize == 512 || sectorSize == 4096 else {
            throw CompoundFileError.unsupportedSectorSize(sectorSize)
        }

        let numberOfFATSectors = Int(data.xlsUInt32(at: 44))
        let firstDIFATSector = Int(data.xlsUInt32(at: 68))
        let numberOfDIFATSectors = Int(data.xlsUInt32(at: 72))
        let difat = try Self.readDIFAT(
            data: data,
            sectorSize: sectorSize,
            headerFATSectorCount: numberOfFATSectors,
            firstDIFATSector: firstDIFATSector,
            numberOfDIFATSectors: numberOfDIFATSectors
        )

        self.fat = try Self.readFAT(data: data, sectorSize: sectorSize, fatSectorIDs: difat)

        let directoryData = try Self.readRegularChain(
            data: data,
            sectorSize: sectorSize,
            fat: fat,
            startSector: firstDirectorySector
        )
        self.directoryEntries = Self.parseDirectoryEntries(from: directoryData)
        self.rootEntry = directoryEntries.first { $0.type == 5 }

        if firstMiniFATSector >= 0, numberOfMiniFATSectors > 0 {
            let miniFATData = try Self.readRegularChain(
                data: data,
                sectorSize: sectorSize,
                fat: fat,
                startSector: firstMiniFATSector,
                sectorLimit: numberOfMiniFATSectors
            )
            self.miniFAT = Self.readUInt32Table(from: miniFATData)
        } else {
            self.miniFAT = []
        }

        if let rootEntry, rootEntry.startingSector >= 0, rootEntry.streamSize > 0 {
            self.miniStream = try Self.readRegularChain(
                data: data,
                sectorSize: sectorSize,
                fat: fat,
                startSector: rootEntry.startingSector,
                byteLimit: rootEntry.streamSize
            )
        } else {
            self.miniStream = Data()
        }
    }

    func stream(named candidateNames: [String]) throws -> Data {
        for name in candidateNames {
            if let entry = directoryEntries.first(where: { $0.type == 2 && $0.name.caseInsensitiveCompare(name) == .orderedSame }) {
                return try stream(for: entry)
            }
        }
        throw CompoundFileError.streamNotFound(candidateNames)
    }

    private func stream(for entry: DirectoryEntry) throws -> Data {
        guard entry.startingSector >= 0, entry.streamSize > 0 else {
            return Data()
        }

        if entry.streamSize < miniStreamCutoffSize, !miniFAT.isEmpty, !miniStream.isEmpty {
            return try Self.readMiniChain(
                miniStream: miniStream,
                miniSectorSize: miniSectorSize,
                miniFAT: miniFAT,
                startSector: entry.startingSector,
                byteLimit: entry.streamSize
            )
        }

        return try Self.readRegularChain(
            data: data,
            sectorSize: sectorSize,
            fat: fat,
            startSector: entry.startingSector,
            byteLimit: entry.streamSize
        )
    }

    private static func readDIFAT(
        data: Data,
        sectorSize: Int,
        headerFATSectorCount: Int,
        firstDIFATSector: Int,
        numberOfDIFATSectors: Int
    ) throws -> [Int] {
        var sectorIDs: [Int] = []

        for index in 0..<109 {
            let value = data.xlsUInt32(at: 76 + index * 4)
            if value != freeSector {
                sectorIDs.append(Int(value))
            }
        }

        var current = firstDIFATSector
        for _ in 0..<numberOfDIFATSectors where current >= 0 {
            let sector = try sectorData(in: data, sectorSize: sectorSize, sectorID: current)
            let entriesPerDIFATSector = sectorSize / 4 - 1
            for index in 0..<entriesPerDIFATSector {
                let value = sector.xlsUInt32(at: index * 4)
                if value != freeSector {
                    sectorIDs.append(Int(value))
                }
            }
            let next = sector.xlsUInt32(at: entriesPerDIFATSector * 4)
            current = next == endOfChain ? -1 : Int(next)
        }

        if sectorIDs.count > headerFATSectorCount {
            return Array(sectorIDs.prefix(headerFATSectorCount))
        }
        return sectorIDs
    }

    private static func readFAT(data: Data, sectorSize: Int, fatSectorIDs: [Int]) throws -> [UInt32] {
        var result: [UInt32] = []
        for sectorID in fatSectorIDs {
            let sector = try sectorData(in: data, sectorSize: sectorSize, sectorID: sectorID)
            result.append(contentsOf: readUInt32Table(from: sector))
        }
        return result
    }

    private static func readUInt32Table(from data: Data) -> [UInt32] {
        stride(from: 0, to: data.count - data.count % 4, by: 4).map {
            data.xlsUInt32(at: $0)
        }
    }

    private static func readRegularChain(
        data: Data,
        sectorSize: Int,
        fat: [UInt32],
        startSector: Int,
        sectorLimit: Int? = nil,
        byteLimit: Int? = nil
    ) throws -> Data {
        var result = Data()
        var seen: Set<Int> = []
        var current = startSector
        var sectorCount = 0

        while current >= 0 {
            if seen.contains(current) {
                throw CompoundFileError.chainLoop(current)
            }
            seen.insert(current)

            let sector = try sectorData(in: data, sectorSize: sectorSize, sectorID: current)
            result.append(sector)
            sectorCount += 1

            if let sectorLimit, sectorCount >= sectorLimit {
                break
            }
            guard current < fat.count else {
                throw CompoundFileError.sectorOutOfBounds(current)
            }

            let next = fat[current]
            if next == endOfChain || next == freeSector || next == fatSector || next == difatSector {
                break
            }
            current = Int(next)
        }

        if let byteLimit, result.count > byteLimit {
            result.removeSubrange(byteLimit..<result.count)
        }
        return result
    }

    private static func readMiniChain(
        miniStream: Data,
        miniSectorSize: Int,
        miniFAT: [UInt32],
        startSector: Int,
        byteLimit: Int
    ) throws -> Data {
        var result = Data()
        var seen: Set<Int> = []
        var current = startSector

        while current >= 0 {
            if seen.contains(current) {
                throw CompoundFileError.chainLoop(current)
            }
            seen.insert(current)

            let offset = current * miniSectorSize
            guard offset >= 0, offset < miniStream.count else {
                throw CompoundFileError.sectorOutOfBounds(current)
            }
            result.append(Data(miniStream[offset..<min(offset + miniSectorSize, miniStream.count)]))

            guard current < miniFAT.count else {
                throw CompoundFileError.sectorOutOfBounds(current)
            }
            let next = miniFAT[current]
            if next == endOfChain || next == freeSector {
                break
            }
            current = Int(next)
        }

        if result.count > byteLimit {
            result.removeSubrange(byteLimit..<result.count)
        }
        return result
    }

    private static func sectorData(in data: Data, sectorSize: Int, sectorID: Int) throws -> Data {
        let offset = (sectorID + 1) * sectorSize
        guard sectorID >= 0, offset >= 0, offset + sectorSize <= data.count else {
            throw CompoundFileError.sectorOutOfBounds(sectorID)
        }
        return Data(data[offset..<offset + sectorSize])
    }

    private static func parseDirectoryEntries(from data: Data) -> [DirectoryEntry] {
        stride(from: 0, to: data.count - data.count % 128, by: 128).compactMap { offset in
            let nameLength = Int(data.xlsUInt16(at: offset + 64))
            guard nameLength >= 2, offset + nameLength <= data.count else { return nil }

            let nameBytes = Data(data[(offset)..<(offset + nameLength - 2)])
            let name = String(data: nameBytes, encoding: .utf16LittleEndian) ?? ""
            let type = data[offset + 66]
            let startingSectorValue = data.xlsUInt32(at: offset + 116)
            let streamSizeLow = UInt64(data.xlsUInt32(at: offset + 120))
            let streamSizeHigh = UInt64(data.xlsUInt32(at: offset + 124))
            let streamSize = Int((streamSizeHigh << 32) | streamSizeLow)

            return DirectoryEntry(
                name: name,
                type: type,
                startingSector: startingSectorValue == endOfChain ? -1 : Int(startingSectorValue),
                streamSize: streamSize
            )
        }
    }
}

extension Data {
    func xlsUInt16(at offset: Int) -> UInt16 {
        guard offset + 2 <= count else { return 0 }
        return withUnsafeBytes { $0.loadUnaligned(fromByteOffset: offset, as: UInt16.self) }.littleEndian
    }

    func xlsUInt32(at offset: Int) -> UInt32 {
        guard offset + 4 <= count else { return 0 }
        return withUnsafeBytes { $0.loadUnaligned(fromByteOffset: offset, as: UInt32.self) }.littleEndian
    }

    func xlsInt32(at offset: Int) -> Int32 {
        Int32(bitPattern: xlsUInt32(at: offset))
    }

    func xlsDouble(at offset: Int) -> Double {
        guard offset + 8 <= count else { return 0 }
        let bitPattern = withUnsafeBytes { $0.loadUnaligned(fromByteOffset: offset, as: UInt64.self) }.littleEndian
        return Double(bitPattern: bitPattern)
    }
}

File: cpp/core/src/compound_file_reader.cpp
#include "compound_file_reader.hpp"

#include <QSet>
#include <QtEndian>
#include <algorithm>
#include <optional>
#include <stdexcept>

namespace xlsone {

namespace {

quint16 u16(const QByteArray& data, qsizetype offset)
{
    if (offset < 0 || offset + 2 > data.size()) {
        return 0;
    }
    return qFromLittleEndian<quint16>(reinterpret_cast<const uchar*>(data.constData() + offset));
}

quint32 u32(const QByteArray& data, qsizetype offset)
{
    if (offset < 0 || offset + 4 > data.size()) {
        return 0;
    }
    return qFromLittleEndian<quint32>(reinterpret_cast<const uchar*>(data.constData() + offset));
}

int sectorIDFromUInt32(quint32 value)
{
    return value == 0xfffffffe ? -1 : static_cast<int>(value);
}

} // namespace

CompoundFileReader::CompoundFileReader(QByteArray data) : data_(std::move(data))
{
    static const QByteArray magic = QByteArray::fromHex("d0cf11e0a1b11ae1");
    if (data_.size() < 512 || data_.left(8) != magic) {
        throw std::runtime_error("不是有效的 OLE Compound File");
    }

    const int sectorShift = u16(data_, 30);
    const int miniSectorShift = u16(data_, 32);
    sectorSize_ = 1 << sectorShift;
    miniSectorSize_ = 1 << miniSectorShift;
    firstDirectorySector_ = sectorIDFromUInt32(u32(data_, 48));
    miniStreamCutoffSize_ = static_cast<int>(u32(data_, 56));
    firstMiniFATSector_ = sectorIDFromUInt32(u32(data_, 60));
    numberOfMiniFATSectors_ = static_cast<int>(u32(data_, 64));

    if (sectorSize_ != 512 && sectorSize_ != 4096) {
        throw std::runtime_error(QStringLiteral("暂不支持的 OLE sector size: %1").arg(sectorSize_).toStdString());
    }

    const int numberOfFATSectors = static_cast<int>(u32(data_, 44));
    const int firstDIFATSector = sectorIDFromUInt32(u32(data_, 68));
    const int numberOfDIFATSectors = static_cast<int>(u32(data_, 72));

    const auto difat = readDIFAT(data_, sectorSize_, numberOfFATSectors, firstDIFATSector, numberOfDIFATSectors);
    fat_ = readFAT(data_, sectorSize_, difat);

    const auto directoryData = readRegularChain(data_, sectorSize_, fat_, firstDirectorySector_);
    directoryEntries_ = parseDirectoryEntries(directoryData);

    if (firstMiniFATSector_ >= 0 && numberOfMiniFATSectors_ > 0) {
        const auto miniFATData = readRegularChain(
            data_,
            sectorSize_,
            fat_,
            firstMiniFATSector_,
            numberOfMiniFATSectors_
        );
        miniFAT_ = readUInt32Table(miniFATData);
    }

    const auto root = std::find_if(directoryEntries_.begin(), directoryEntries_.end(), [](const DirectoryEntry& entry) {
        return entry.type == 5;
    });
    if (root != directoryEntries_.end() && root->startingSector >= 0 && root->streamSize > 0) {
        miniStream_ = readRegularChain(data_, sectorSize_, fat_, root->startingSector, std::nullopt, root->streamSize);
    }
}

QByteArray CompoundFileReader::stream(const QStringList& candidateNames) const
{
    for (const auto& name : candidateNames) {
        const auto iterator = std::find_if(directoryEntries_.begin(), directoryEntries_.end(), [&](const DirectoryEntry& entry) {
            return entry.type == 2 && entry.name.compare(name, Qt::CaseInsensitive) == 0;
        });
        if (iterator != directoryEntries_.end()) {
            return streamFor(*iterator);
        }
    }
    throw std::runtime_error(QStringLiteral("未找到工作簿数据流: %1").arg(candidateNames.join(QStringLiteral(", "))).toStdString());
}

QByteArray CompoundFileReader::streamFor(const DirectoryEntry& entry) const
{
    if (entry.startingSector < 0 || entry.streamSize <= 0) {
        return {};
    }
    if (entry.streamSize < miniStreamCutoffSize_ && !miniFAT_.empty() && !miniStream_.isEmpty()) {
        return readMiniChain(miniStream_, miniSectorSize_, miniFAT_, entry.startingSector, entry.streamSize);
    }
    return readRegularChain(data_, sectorSize_, fat_, entry.startingSector, std::nullopt, entry.streamSize);
}

std::vector<int> CompoundFileReader::readDIFAT(
    const QByteArray& data,
    int sectorSize,
    int headerFATSectorCount,
    int firstDIFATSector,
    int numberOfDIFATSectors
)
{
    std::vector<int> sectorIDs;
    for (int index = 0; index < 109; ++index) {
        const quint32 value = u32(data, 76 + index * 4);
        if (value != FreeSector) {
            sectorIDs.push_back(static_cast<int>(value));
        }
    }

    int current = firstDIFATSector;
    for (int count = 0; count < numberOfDIFATSectors && current >= 0; ++count) {
        const auto sector = sectorData(data, sectorSize, current);
        const int entriesPerDIFATSector = sectorSize / 4 - 1;
        for (int index = 0; index < entriesPerDIFATSector; ++index) {
            const quint32 value = u32(sector, index * 4);
            if (value != FreeSector) {
                sectorIDs.push_back(static_cast<int>(value));
            }
        }
        const quint32 next = u32(sector, entriesPerDIFATSector * 4);
        current = next == EndOfChain ? -1 : static_cast<int>(next);
    }

    if (sectorIDs.size() > static_cast<size_t>(headerFATSectorCount)) {
        sectorIDs.resize(static_cast<size_t>(headerFATSectorCount));
    }
    return sectorIDs;
}

std::vector<quint32> CompoundFileReader::readFAT(const QByteArray& data, int sectorSize, const std::vector<int>& fatSectorIDs)
{
    std::vector<quint32> result;
    for (const int sectorID : fatSectorIDs) {
        const auto sector = sectorData(data, sectorSize, sectorID);
        const auto values = readUInt32Table(sector);
        result.insert(result.end(), values.begin(), values.end());
    }
    return result;
}

std::vector<quint32> CompoundFileReader::readUInt32Table(const QByteArray& data)
{
    std::vector<quint32> result;
    for (qsizetype offset = 0; offset + 4 <= data.size(); offset += 4) {
        result.push_back(u32(data, offset));
    }
    return result;
}

QByteArray CompoundFileReader::readRegularChain(
    const QByteArray& data,
    int sectorSize,
    const std::vector<quint32>& fat,
    int startSector,
    std::optional<int> sectorLimit,
    std::optional<qsizetype> byteLimit
)
{
    QByteArray result;
    QSet<int> seen;
    int current = startSector;
    int sectorCount = 0;

    while (current >= 0) {
        if (seen.contains(current)) {
            throw std::runtime_error(QStringLiteral("OLE sector chain 出现循环: %1").arg(current).toStdString());
        }
        seen.insert(current);

        result.append(sectorData(data, sectorSize, current));
        ++sectorCount;

        if (sectorLimit.has_value() && sectorCount >= *sectorLimit) {
            break;
        }
        if (current >= static_cast<int>(fat.size())) {
            throw std::runtime_error(QStringLiteral("OLE sector 越界: %1").arg(current).toStdString());
        }

        const quint32 next = fat[static_cast<size_t>(current)];
        if (next == EndOfChain || next == FreeSector || next == FatSector || next == DifatSector) {
            break;
        }
        current = static_cast<int>(next);
    }

    if (byteLimit.has_value() && result.size() > *byteLimit) {
        result.truncate(*byteLimit);
    }
    return result;
}

QByteArray CompoundFileReader::readMiniChain(
    const QByteArray& miniStream,
    int miniSectorSize,
    const std::vector<quint32>& miniFAT,
    int startSector,
    qsizetype byteLimit
)
{
    QByteArray result;
    QSet<int> seen;
    int current = startSector;

    while (current >= 0) {
        if (seen.contains(current)) {
            throw std::runtime_error(QStringLiteral("OLE mini sector chain 出现循环: %1").arg(current).toStdString());
        }
        seen.insert(current);

        const qsizetype offset = current * miniSectorSize;
        if (offset < 0 || offset >= miniStream.size()) {
            throw std::runtime_error(QStringLiteral("OLE mini sector 越界: %1").arg(current).toStdString());
        }
        result.append(miniStream.mid(offset, std::min<qsizetype>(miniSectorSize, miniStream.size() - offset)));

        if (current >= static_cast<int>(miniFAT.size())) {
            throw std::runtime_error(QStringLiteral("OLE mini FAT 越界: %1").arg(current).toStdString());
        }
        const quint32 next = miniFAT[static_cast<size_t>(current)];
        if (next == EndOfChain || next == FreeSector) {
            break;
        }
        current = static_cast<int>(next);
    }

    if (result.size() > byteLimit) {
        result.truncate(byteLimit);
    }
    return result;
}

QByteArray CompoundFileReader::sectorData(const QByteArray& data, int sectorSize, int sectorID)
{
    const qsizetype offset = (sectorID + 1) * sectorSize;
    if (sectorID < 0 || offset < 0 || offset + sectorSize > data.size()) {
        throw std::runtime_error(QStringLiteral("OLE sector 越界: %1").arg(sectorID).toStdString());
    }
    return data.mid(offset, sectorSize);
}

std::vector<CompoundFileReader::DirectoryEntry> CompoundFileReader::parseDirectoryEntries(const QByteArray& data)
{
    std::vector<DirectoryEntry> entries;
    for (qsizetype offset = 0; offset + 128 <= data.size(); offset += 128) {
        const int nameLength = u16(data, offset + 64);
        if (nameLength < 2 || offset + nameLength > data.size()) {
            continue;
        }

        const QByteArray nameBytes = data.mid(offset, nameLength - 2);
        const QString name = QString::fromUtf16(reinterpret_cast<const char16_t*>(nameBytes.constData()), nameBytes.size() / 2);
        const quint32 startingSector = u32(data, offset + 116);
        const quint64 streamSize = static_cast<quint64>(u32(data, offset + 120))
            | (static_cast<quint64>(u32(data, offset + 124)) << 32);

        entries.push_back({
            name,
            static_cast<quint8>(data[offset + 66]),
            startingSector == EndOfChain ? -1 : static_cast<int>(startingSector),
            static_cast<qsizetype>(streamSize)
        });
    }
    return entries;
}

} // namespace xlsone
